istioctl/cmd/revision.go

```diff
@@ -0,0 +1,659 @@
+	revisionCmd := &cobra.Command{
+		Use:     "revision",
+		Short:   "Revision centric view of Istio deployment",
+		Aliases: []string{"rev"},
+	}
+	revisionCmd.PersistentFlags().StringVarP(&revArgs.manifestsPath, "manifests", "d", "", mesh.ManifestsFlagHelpStr)
+	revisionCmd.PersistentFlags().BoolVarP(&revArgs.verbose, "verbose", "v", false, "print customizations")
+
+	revisionCmd.AddCommand(revisionListCommand())
+	revisionCmd.AddCommand(revisionDescribeCommand())
+	return revisionCmd
+}
+
+// TODO(su225): Fix example, description - command documentation part
+func revisionDescribeCommand() *cobra.Command {
+	describeCmd := &cobra.Command{
+		Use:     "describe",
+		Example: `    istioctl experimental revision describe <revision>`,
+		Short:   "Show details of a revision - customizations, number of pods pointing to it, istiod, gateways etc",
+		Args: func(cmd *cobra.Command, args []string) error {
+			if len(args) == 0 {
+				return fmt.Errorf("revision is not specified")
+			}
+			if len(args) > 1 {
+				return fmt.Errorf("exactly 1 revision should be specified")
+			}
+			if revArgs.verbose && len(revArgs.sections) != 0 {
+				return fmt.Errorf("sections and verbose flags should not be specified at the same time")
+			}
+			isValidFormat := false
+			for _, f := range validFormats {
+				if revArgs.output == f {
+					isValidFormat = true
+					break
+				}
+			}
+			if !isValidFormat {
+				return fmt.Errorf("unknown format %s. It should be %#v", revArgs.output, validFormats)
+			}
+			return nil
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			revArgs.name = args[0]
+			if revArgs.name != "<default>" {
+				if errs := validation.IsDNS1123Label(revArgs.name); len(errs) > 0 {
+					return fmt.Errorf(strings.Join(errs, "\n"))
+				}
+			} else {
+				revArgs.name = "" // TODO(su225): This case is broken. Fix it.
+			}
+			logger := clog.NewConsoleLogger(cmd.OutOrStdout(), cmd.ErrOrStderr(), scope)
+			return revisionDescription(cmd.OutOrStdout(), &revArgs, logger)
+		},
+	}
+	describeCmd.Flags().BoolVarP(&revArgs.verbose, "verbose", "v", false, "Dump all information related to the revision")
+	describeCmd.Flags().StringSliceVarP(&revArgs.sections, "sections", "s", []string{}, "Sections that should be included in description output")
+	describeCmd.Flags().StringVarP(&revArgs.output, "output", "o", "table", "Output format for revision description")
+	return describeCmd
+}
+
+func revisionListCommand() *cobra.Command {
+	listCmd := &cobra.Command{
+		Use:     "list",
+		Short:   "Show list of control plane and gateway revisions that are currently installed in cluster",
+		Example: `   istioctl experimental revision list`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			logger := clog.NewConsoleLogger(cmd.OutOrStdout(), cmd.ErrOrStderr(), scope)
+			return revisionList(cmd.OutOrStdout(), &revArgs, logger)
+		},
+	}
+	return listCmd
+}
+
+// TODO: Come up with a better name
+type revisionInfo struct {
```

Comment:
How about naming it?
```suggestion
type revisionConfig struct {
```

Reply:
Yeah. It sounds good. Renamed it.